specs/Slots.md

```diff
@@ -0,0 +1,64 @@
+# Slots API specification
+
+Fluent UI components almost always contain sub parts and these pre-defined layouts can be configured via Slots API. Slots are named areas in a component that can receive props and provide styling and layout for them.
+
+## Slots definition
+
+For example, `Button` component contains `content`, `loader` and `icon` slots in its layout:
+
+```jsx
+<root>
+  <loader />
+  <content />
+  <icon />
+</root>
+```
+
+## Slots usage
+
+Slots can be configured through props, which lets the caller pass in a variety of inputs for a given slot. There are several forms of values that can be provided, but all of them share one common thing - they will produce a proper JSX layout inside a component.
+
+### An object as a value
+
+If you will pass an object to component props we will handle it as props for a slot. In an example below, props will be passed to an `icon` slot:
+
+```jsx
+<Button icon={{ children: <FooIcon />, id: '#button-icon' }} />
+```
+
+### Primitives as a value
+
+We consider JSX elements, strings and numbers as primitive values that will be passed to a meaningful prop, usually to `children`:
+
+```jsx
+<>
+  <Button icon={<FooIcon />} />
+  {/* 💡 has identical effect to the previous one */}
+  <Button icon={{ children: <FooIcon /> }} />
```

Comment:
Is the second example syntactic sugar for the first ?

Reply:
or should they work completely differently under the hood ?

Reply:
No need for tech implem details, but we should be clear if one is an alias for the other

Reply:
I think the first example is just a shorthand for the more explicit second example according to [react-examples/src/compose/getSlots.ts#L33](https://github.com/microsoft/fluentui/blob/master/packages/react-utils/src/compose/getSlots.ts#L33). Ofc please do correct me if I'm wrong!

Reply:
So I looked briefly at current impl of `getSlots` and this statement is misleading (correct me if I'm wrong). 

From JSX transpilation perspective:

`<Button icon={<FooIcon />} />` -> `<Button icon={React.createElement(FooIcon)} />`  ->`<Button icon={{ type: FooIcon, children: null }} />`

how can that be identical to `<Button icon={{ children: <FooIcon /> }} />` ? Is the implementation gonna change ?

Reply:
OK, so I just checked and you're right guys, implementation wise, the two are different:
1. generates 

```
Button props:
icon: <FooIcon />
```

2. generates

```
Button props:
icon: { children: <FooIcon /> }
```

When rendered in the DOM, however, they do come down to the same set of elements:

```html
<button ...>
    <span class="... icon-"
         <!-- whatever FooIcon is rendered as -->
    </span>
</button>
```

Reply:
@kubkon @ling1726 @Hotell thanks for feedback 👍 

I updated this block to (text around it was also updated):

```jsx
<>
  <Button icon={{ children: <FooIcon /> }} />
  {/* 💡 will produce the same JSX/HTML markup, as will be expanded to { children: <FooIcon /> } */}
  <Button icon={<FooIcon />} />
</>
```

(hope that now it has more sense)

We have a special handling for JSX elements, strings and number, so they will be expanded to `{ children: VALUE }`. In code it's done in `resolveShorthands()`:

https://github.com/microsoft/fluentui/blob/661d1d82d79e606a7fa77ed1125ffc6528325ec9/packages/react-utilities/src/compose/resolveShorthandProps.tsx#L22-L25